internal/config/attributemap_test.go

```diff
@@ -0,0 +1,355 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAttributeMapPopBool(t *testing.T) {
+	tests := []struct {
+		desc     string
+		give     AttributeMap
+		giveName string
+
+		want       bool
+		wantErrors []string
+	}{
+		{
+			desc: "true decode",
+			give: AttributeMap{
+				"true": true,
+			},
+			giveName: "true",
+			want:     true,
+		},
+		{
+			desc: "false decode",
+			give: AttributeMap{
+				"false": false,
+			},
+			giveName: "false",
+			want:     false,
+		},
+		{
+			desc: "invalid decode",
+			give: AttributeMap{
+				"test": "teeeeeest",
+			},
+			giveName:   "test",
+			wantErrors: []string{"failed to read attribute", `"test"`, `teeeeeest`},
```

Comment:
biggest reason why it is done is that the errors here are not concrete types we can assert against.  They are a combination of private parsing errors from mapdecode, multierrs that combine multiple error messages, and our own defined Errorfs.  If I changed this to test the error type directly, It would look almost the same, just put into a struct... :man_shrugging: